jenkins/bootstrap_test.py

```diff
@@ -0,0 +1,460 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for bootstrap."""
+
+import json
+import os
+import subprocess
+import unittest
+
+import bootstrap
+
+
+BRANCH = 'random_branch'
+FAIL = ['/bin/bash', '-c', 'exit 1']
+JOB = 'random_job'
+PASS = ['/bin/bash', '-c', 'exit 0']
+PULL = 12345
+REPO = 'random_org/random_repo'
+
+
+class Stub(object):
+    """Replace thing.param with replacement until exiting with."""
+    def __init__(self, thing, param, replacement):
+        self.thing = thing
+        self.param = param
+        self.replacement = replacement
+        self.old = getattr(thing, param)
+        setattr(thing, param, self.replacement)
+
+    def __enter__(self, *a, **kw):
+        return self.replacement
+
+    def __exit__(self, *a, **kw):
+        setattr(self.thing, self.param, self.old)
+
+
+class FakeSubprocess(object):
+    """Keep track of calls."""
+    def __init__(self):
+        self.calls = []
+
+    def __call__(self, cmd, *a, **kw):
+        self.calls.append((cmd, a, kw))
+
+
+def Pass(*a, **kw):
+    """Do nothing."""
+    pass
+
+
+def Truth(*a, **kw):
+    """Always true."""
+    return True
+
+
+def Bomb(*a, **kw):
+    """Always raise."""
+    raise AssertionError('Should not happen', a, kw)
+
+
+class SubprocessTest(unittest.TestCase):
+    """Tests for Subprocess()."""
+
+    def testStdin(self):
+        """Will write to subprocess.stdin."""
+        with self.assertRaises(subprocess.CalledProcessError) as cpe:
+            bootstrap.Subprocess(['/bin/bash'], stdin='exit 92')
+        self.assertEquals(92, cpe.exception.returncode)
+
+    def testCheckTrue(self):
+        """Raise on non-zero exit codes if check is set."""
+        with self.assertRaises(subprocess.CalledProcessError) as cpe:
+            bootstrap.Subprocess(FAIL, check=True)
+
+        bootstrap.Subprocess(PASS, check=True)
+
+    def testCheckDefault(self):
+        """Default to check=True."""
+        with self.assertRaises(subprocess.CalledProcessError) as cpe:
+            bootstrap.Subprocess(FAIL)
+
+        bootstrap.Subprocess(PASS)
+
+    def testCheckFalse(self):
+        """Never raise when check is not set."""
+        bootstrap.Subprocess(FAIL, check=False)
+        bootstrap.Subprocess(PASS, check=False)
+
+    def testOutput(self):
+        """Output is returned when requested."""
+        cmd = ['/bin/bash', '-c', 'echo hello world']
+        self.assertEquals(
+            'hello world\n', bootstrap.Subprocess(cmd, output=True))
+
+class CheckoutTest(unittest.TestCase):
```

Comment:
Can you make a test that ends up running something like `git clone /tmp/tmp-repo` to verify that our git syntax is correct?